src/GafferOSL/ShadingEngine.cpp

```diff
@@ -1241,6 +1508,222 @@ IECore::CompoundDataPtr ShadingEngine::shade( const IECore::CompoundData *points
+
+	RenderState renderState( points, transforms, contextVariablesNeeded, context );
+
+	// Get pointers to varying data, we'll use these to
+	// update the shaderGlobals as we iterate over our points.
+
+	const float *u = varyingValue<float>( points, "u" );
+	const float *v = varyingValue<float>( points, "v" );
+	const V2f *uv = varyingValue<V2f>( points, "uv" );
+	const V3f *n = varyingValue<V3f>( points, "N" );
+
+	/// \todo Get the other globals - match the uniform list
+
+	// Allocate data for the result
+
+	ShadingResults results( numPoints );
+
+	// Iterate over the input points, doing the shading as we go
+
+	tbb::enumerable_thread_specific<ThreadInfo> threadInfoCache;
+
+	const IECore::Canceller *canceller = context->canceller();
+
+
+	auto f = [&shadingSystem, &renderState, &results, &shaderGlobals, &p, &u, &v, &uv, &n, &shaderGroup, &threadInfoCache, canceller]( const tbb::blocked_range<size_t> &r )
+	{
+		ThreadInfo &threadInfo = threadInfoCache.local();
+		shadingSystem->batched<WidthT>().jit_group( &shaderGroup, threadInfo.shadingContext );
+
+		ThreadRenderState threadRenderState( renderState );
+
+		BatchedShaderGlobals< WidthT > threadShaderGlobals;
+		memset( (void *)&threadShaderGlobals, 0, sizeof( BatchedShaderGlobals<WidthT> ) );
+
+		OSL::assign_all( threadShaderGlobals.varying.time, shaderGlobals.time );
+
+		OSL::assign_all( threadShaderGlobals.varying.dPdx, shaderGlobals.dPdx );
+		OSL::assign_all( threadShaderGlobals.varying.dPdy, shaderGlobals.dPdy );
+		OSL::assign_all( threadShaderGlobals.varying.dPdz, shaderGlobals.dPdz );
+
+		OSL::assign_all( threadShaderGlobals.varying.I, shaderGlobals.I );
+		OSL::assign_all( threadShaderGlobals.varying.dIdx, shaderGlobals.dIdx );
+		OSL::assign_all( threadShaderGlobals.varying.dIdy, shaderGlobals.dIdy );
+
+		OSL::assign_all( threadShaderGlobals.varying.N, shaderGlobals.N );
+		OSL::assign_all( threadShaderGlobals.varying.Ng, shaderGlobals.Ng );
+
+		OSL::assign_all( threadShaderGlobals.varying.u, shaderGlobals.u );
+		OSL::assign_all( threadShaderGlobals.varying.dudx, shaderGlobals.dudx );
+		OSL::assign_all( threadShaderGlobals.varying.dudy, shaderGlobals.dudy );
+
+		OSL::assign_all( threadShaderGlobals.varying.v, shaderGlobals.v );
+		OSL::assign_all( threadShaderGlobals.varying.dvdx, shaderGlobals.dvdx );
+		OSL::assign_all( threadShaderGlobals.varying.dvdy, shaderGlobals.dvdy );
+
+		OSL::assign_all( threadShaderGlobals.varying.dPdu, shaderGlobals.dPdu );
+		OSL::assign_all( threadShaderGlobals.varying.dPdv, shaderGlobals.dPdv );
+
+		threadShaderGlobals.uniform.renderstate = &threadRenderState;
+
+		for( size_t i = r.begin(); i < r.end(); i += WidthT )
+		{
+			int batchSize = std::min( size_t( WidthT ), size_t( r.end() ) - i );
+
+			OSL::Block<int, WidthT> wideShadeIndex;
+			for( int j = 0; j < batchSize; j++ )
+			{
+				wideShadeIndex[j] = i + j;
+				threadShaderGlobals.varying.P[j] = p[i + j];
+			}
+
+			IECore::Canceller::check( canceller );
```

Comment:
Can we move this check back to the start of the block? Conceptually we want to check before doing anything, rather than part way through. And it'll help maintain these two really similar functions in sync if they keep to the same basic structure.

Reply:
Addressed in 8c84600